device/src/region/fixed_channel_plans/mod.rs

```diff
@@ -8,15 +8,71 @@ mod us915;
 pub(crate) use au915::AU915;
 pub(crate) use us915::US915;
 
+#[derive(Clone)]
+struct PreferredJoinChannels {
+    channel_list: heapless::Vec<u8, 16>,
+    // Decrementing number representing how many tries we have left with the specified list before
+    // reverting to the default channel selection behavior.
+    num_retries: usize,
+}
+
+#[derive(Default, Clone)]
+struct JoinChannels {
+    preferred_channels: Option<PreferredJoinChannels>,
+    // List of subbands that haven't already been tried.
+    available_subbands: heapless::Vec<u8, 9>,
+}
+
+impl JoinChannels {
+    fn select_channel(&mut self, rng: &mut impl RngCore) -> u8 {
+        let random = rng.next_u32();
+
+        if let Some(fav) = &mut self.preferred_channels {
+            if fav.num_retries > 0 {
```

Comment:
It looks like this retry logic will only apply to the first Join. It's probably good to not mutate the `num_retries` after construction but set a retry counter for ever Join sequence.

Reply:
What constitutes a join sequence? As far as I can tell, the user is responsible for calling `join()` until the network replies.